crates/sparrow-runtime/src/prepare/column_behavior.rs

```diff
@@ -0,0 +1,522 @@
+                    Error::NullInNonNullableColumn {
+                        field: batch.schema().field(*index).name().to_owned(),
+                        null_count: column.null_count()
+                    }
+                );
+
+                order_preserving_cast_to_u64(column)
+                    .into_report()
+                    .change_context(Error::PreparingColumn)?
+            }
+            ColumnBehavior::Reference { index, nullable } => {
+                let column = batch.column(*index);
+                error_stack::ensure!(
+                    *nullable || column.null_count() == 0,
+                    Error::NullInNonNullableColumn {
+                        field: batch.schema().field(*index).name().to_owned(),
+                        null_count: column.null_count()
+                    }
+                );
+                column.clone()
+            }
+            ColumnBehavior::PrepareEntityKey { index, nullable } => {
+                let metadata = metadata.expect("metadata");
+                let column = batch.column(*index);
+                error_stack::ensure!(
+                    *nullable || column.null_count() == 0,
+                    Error::NullInNonNullableColumn {
+                        field: batch.schema().field(*index).name().to_owned(),
+                        null_count: column.null_count()
+                    }
+                );
+                let previous_keys = metadata.previous_keys.borrow_mut();
+                // 1. Hash the current entity key column to a UInt64Array
+                let entity_column = sparrow_kernels::hash::hash(column)
+                    .into_report()
+                    .change_context(Error::PreparingColumn)?;
+                // 2. Convert the hash column specifically to a UInt64Array from a dyn array.
+                let indices: UInt64Array = entity_column
+                    .iter()
+                    .flatten()
+                    .enumerate()
+                    .filter_map(|(index, key_hash)| {
+                        if previous_keys.insert(key_hash) {
+                            Some(index as u64)
+                        } else {
+                            None
+                        }
+                    })
+                    .collect();
+                let new_hash_keys = arrow::compute::take(&entity_column, &indices, None)
+                    .into_report()
+                    .change_context(Error::PreparingColumn)?;
+                let new_keys = arrow::compute::take(column, &indices, None)
+                    .into_report()
+                    .change_context(Error::PreparingColumn)?;
+                metadata
+                    .add_entity_keys(new_hash_keys, new_keys)
+                    .into_report()
+                    .change_context(Error::PreparingColumn)?;
+                Arc::new(entity_column)
+            }
+            ColumnBehavior::ExecuteEntityKey { index, nullable } => {
```

Comment:
Differentiated from the `PrepareEntityKey` because in prepare we use metadata files to store the key hash inverse mappings, whereas during execution we have to populate it dynamically.